### ISSUE TITLE
Parameters and behaviour of CRUD actions

### ISSUE THREAD

**Issue by JanMalch** [ngxs-labs/entity-state#4]
I like Akitas approach of having few actions and allowing different parameter types, like updating by a single ID, an array of IDs or a selecting by function. This is also the current implementation.

>https://netbasal.gitbook.io/akita/entity-store/entity-store/api

One particular interesting action is `createOrReplace` vs `add`. The former takes an ID and an entity, while the latter only takes one or more entities.

**Comment by JanMalch** [comment]
I would prefer having only one `addOrReplace` that behaves like `add`, so you have to update your entity beforehand with a valid ID.

**Comment by markwhitfeld** [comment]
A code example would be great here... ;-)

**Comment by JanMalch** [comment]
So the actions all use these kind of parameters:

```typescript
// means     EntitySelector<T> = one id | multiple ids | all entities that return true | all entities;
export type EntitySelector<T> = string | string[] | ((T) => boolean) | null;
// means     Updater<T> = Partial entity | custom update function, runs before classes #onUpdate
export type Updater<T> = Partial<T> | ((entity: T) => Partial<T>);
```

Question is: should there just be one `update`, `add` and `remove` that accepts an parameter of this type or multiple different functions like `update, `updateMany`, ...

---

About `addOrReplace`: I'm saying that I don't see the need for Akita's `createOrReplace(id: ID, entity: E)`

**Comment by JanMalch** [comment]
I'm not sure what you mean? `ID` is `string | number`. I currently only use `string` (I think I had some weird issues with `number` once, but I can look into that). 

I currently don't see the difference in `createOrReplace` and `add` apart from the ID as parameter.

**Comment by odahcam** [comment]
Yeah, but my concern is: how will you know what entity should be replaced?

**Comment by odahcam** [comment]
Yeah, but my concern is: how will you know what entity should be replaced?

**Comment by JanMalch** [comment]
I take the ID of the passed entity and just overwrite that value in the entity map.

https://github.com/ngxs-labs/entity-state/blob/master/projects/entity-state/src/lib/entity-state.ts#L174

**Comment by odahcam** [comment]
Understood, users must configure or implement the default `idKey` so the algorithm can check over this field.

**Comment by JanMalch** [comment]
Exactly. You pass the key as string here: https://github.com/ngxs-labs/entity-state/blob/master/src/app/store/todo/store.ts#L17 it has to be `keyof T`

**Comment by odahcam** [comment]
Well, I think `add` and `update` could be used where `update` falls to `add` if the entity doesn't exist, or something like that. Also, not everyone will want this fallback and some will just prefer using update instead of add everytime. I think we should have the options here.

**Comment by JanMalch** [comment]
I'm not talking about `add` vs `update`. The consquences are also in the name: `addOrReplace`. `update` will throw an error if the entity doesn't exist: https://github.com/ngxs-labs/entity-state/blob/master/projects/entity-state/src/lib/entity-state.ts#L272

I'm talking about Akitas `createOrReplace` vs their `add`. The former takes ID as a parameter while `add` just takes an entity. They also have an `upsert` action. I'm not exactly sure what the differences are, if there are different side effects.

What I'm asking is, is: Should there be additional actions to maybe differentiate with ActionHandlers or similar. I don't see the need as it's confusing what the actual differences are. 

So my opinion is: Just have two Actions `addOrReplace` and `update`. The latter one fails if the entity doesn't exist. If you don't want to run in this error use the `replace` in `addOrReplace`.

**Comment by odahcam** [comment]
I may want to add the same item twice, don't ask me why, users **will** try to do this, so I think a `add` is needed too.

**Comment by JanMalch** [comment]
How do you add two rows with the same primary key in a SQL database?
And who do you mean by the user? The website user or the developer? It's the developers job to make sure that there are unique IDs. The library could add a UUID function or something but I wouldn't want my ngxs store to generate some ID automatically without me knowing how it did this. Or that it did this. In the end the data has to match my schema on the server aswell.

**Comment by odahcam** [comment]
> And who do you mean by the user? The website user or the developer? 

In this case the user is the developer, we're on the library/package scope.

> How do you add two rows with the same primary key in a SQL database?

It is not the states function to mimic a SQL database, it should just store data. 

Sometimes users will just register input into the state and they want that some items just repeat, think in a scanner for example, it will read and generate lots of results, some identical to others. In this case they don't even need an id, but you can also think about a game where the user store items in a bag, etc. I think we should not prevent this to happen, because each approach has its benefits.

**Comment by JanMalch** [comment]
So you propse an array as the state collection? How would you remove a certain entity?
So your state would look something like this: `[{ title: "foo.pdf" }, { title: "bar.pdf" }]`. Why not put `{ "foo.pdf": { title: "foo.pdf" }, "bar.pdf": { title: "bar.pdf" } }` in your state?
And if you have the same two entities `[{ title: "foo.pdf" }, { title: "foo.pdf" }]`, how do you determine behaviour? Maybe I expected to not have the same entity twice. I mean they are identical really. And what do you expect on a remove? Remove them both? Or just one? If I load my array again, how do I update just one of them? Also how do I choose between Array and HashMap implementation? Do I extend a different class? How should an unexperienced developer now what to choose, if he doesn't even understand hash maps? 
I really don't see any use case for this. It would make things just way more confusing and have unexpected behaviour.

**Comment by odahcam** [comment]
Well, an auto-incrementable key could be used. I'm just saying that is good to have the option, but maybe this could be solved with a ordinary state with a array property and the option would make no sense.

About the `add`, what if this method throws an Error if the entity the user is trying to add is already present?

**Comment by JanMalch** [comment]
There is no error, because there is no `add`. It's called `addOrReplace`. Maybe my original question was badly phrased.

So I would say I always want to have an ID in my implementation of this library. NoSQL has document ID afaik and imo entity states do want to mimic databases somewhat, because otherwise I can just implement my own state class.

Anyways, I would propose the following:

- in the `super` call a third parameter is added: `idStrategy`. It's an enum with 3 options: `fromEntity`, `increment` and `uuid`.
- new action `add(Entity)`: with fromEntity strategy it will throw an error if the ID-value is undefined. If it already exists, it will throw an error. In all cases it will do `entity[idKey] = id;` (where `id` is determined by strategy), which is important for uuid and incr. If I have a list of entities it's necessary to have the ID present, right? End result will be the same for all strategies: a new entity (unless there was an error) will be added.
- new action `createOrReplace(Entity)`: similar to `add` but it will check if a valid ID is in the entity beforehand. If not it will generate one for the automatic idStrategies. So you can use that on a new-or-edit page. End result will be the same for all strategies: if a valid ID is present or can be generated it will use that and create/replace without error.

Please tell me if I missed a use-case.

**Comment by odahcam** [comment]
I think these actions will cover any add case.

**Comment by JanMalch** [comment]
Implemented idGenerators in  6beb0454304a7f258407539ab6ff4d42fde32b73
Put to use in 0006457382c12bd0153ed1dab192d4571640d76c

**Comment by JanMalch** [comment]
Closing this as the syntax is set for now.
